packages/fx-core/src/core/middleware/MigrationUtils.ts

```diff
@@ -111,3 +112,162 @@ export async function needMigrateToAadManifest(ctx: CoreHookContext): Promise<bo
+  "state.fx-resource-bot.siteEndpoint",
+  "state.fx-resource-identity.identityName",
+  "state.fx-resource-identity.identityResourceId",
+  "state.fx-resource-identity.identityClientId",
+  "state.fx-resource-function.sku",
+  "state.fx-resource-function.appName",
+  "state.fx-resource-function.domain",
+  "state.fx-resource-function.appServicePlanName",
+  "state.fx-resource-function.functionAppResourceId",
+  "state.fx-resource-function.functionEndpoint",
+  "state.fx-resource-key-vault.m365ClientSecretReference",
+  "state.fx-resource-key-vault.botClientSecretReference",
+];
+export const pluginIdMappingV3: { [key: string]: string } = {
+  "fx-resource-frontend-hosting": "teams-tab",
+  "fx-resource-function": "teams-api",
+  "fx-resource-identity": "identity",
+  "fx-resource-bot": "teams-bot",
+  "fx-resource-key-vault": "key-vault",
+  "fx-resource-azure-sql": "azure-sql",
+  "fx-resource-apim": "apim",
+};
+const secretPrefix = "SECRET_";
+const configPrefix = "CONFIG__";
+const provisionOutputPrefix = "PROVISIONOUTPUT__";
+
+function generateOutputNameRegexForPlugin(pluginId: string) {
+  return new RegExp(
+    "output +(\\S+) +object += +{" + // Mataches start of output declaration and capture output name. Example: output functionOutput object = {
+      "[^{]*" + // Matches everything between '{' and plugin id declaration. For example: comments, extra properties. Will match multilines.
+      `teamsFxPluginId: +\'${pluginId}\'`, // Matches given plugin id declaration
+    "g"
+  );
+}
+
+export function namingConverterV3(
+  name: string,
+  type: FileType,
+  bicepContent: string
+): Result<string, FxError> {
+  try {
+    if (Object.keys(fixedNamingsV3).includes(name)) {
+      return ok(fixedNamingsV3[name]);
+    } else if (
+      provisionOutputNamingsV3.some((element, index, array) => {
+        // for sql, may have key like: state.fx-resource-azure-sql.databaseName_xxx
+        return name.startsWith(element);
+      })
+    ) {
+      const res = provisionOutputNamingConverterV3(name, bicepContent);
+      return ok(res);
+    } else {
+      const res = commonNamingConverterV3(name);
+      switch (type) {
+        case FileType.CONFIG:
+          return ok(`${configPrefix}${res}`);
+        case FileType.USERDATA:
+          return ok(`${secretPrefix}${res}`);
+        case FileType.STATE:
+        default:
+          return ok(res);
+      }
+    }
+  } catch (error: any) {
+    return err(new SystemError(CoreSource, "FailedToConvertV2ConfigNameToV3", error?.message));
+  }
+}
+
+// convert x-xx.xxx.xxx to x_xx__xxx__xxx
+function commonNamingConverterV3(name: string): string {
+  const names = name.split(".");
+  return names.join("__").replace(/\-/g, "_").toUpperCase();
+}
+
+function provisionOutputNamingConverterV3(name: string, bicepContent: string): string {
+  const names = name.split(".");
+  const pluginNames = [names[1], pluginIdMappingV3[names[1]]];
+  const keyName = names[2];
```

Comment:
Just comfirm that all the env names used in V2 will be no more than 2 layers nested(2 "." at most). Right?

Reply:
For these special cases, we have a list above and all the items in the list are in the same format.